libraries/chem-meta/src/parsing-utils/chemical-table-parser.ts

```diff
@@ -0,0 +1,196 @@
+
+  public init(file: string): void {
+    this.file = file.replaceAll('\r', '');
+    this._atomCount = undefined;
+    this._atomTypes = undefined;
+    this._bondCount = undefined;
+    this.atomCoordinates = undefined;
+  }
+
+  get atomCount(): number {
+    if (this._atomCount === undefined)
+      this.setAtomAndBondCounts();
+    return this._atomCount!;
+  }
+
+  /** Number of bonds in a molecule  */
+  get bondCount(): number {
+    if (this._bondCount === undefined)
+      this.setAtomAndBondCounts();
+    return this._bondCount!;
+  }
+
+  /** X coordinates of all atoms in a molecule  */
+  get x(): Float32Array {
+    if (this.atomCoordinates === undefined)
+      this.atomCoordinates = this.parseAtomCoordinates();
+    return this.atomCoordinates.x;
+  };
+
+  /** Y coordinates of all atoms in a molecule  */
+  get y(): Float32Array {
+    if (this.atomCoordinates === undefined)
+      this.atomCoordinates = this.parseAtomCoordinates();
+    return this.atomCoordinates!.y;
+  };
+
+  /** Z coordinates of all atoms in a molecule  */
+  get z(): Float32Array {
+    if (this.atomCoordinates === undefined)
+      this.atomCoordinates = this.parseAtomCoordinates();
+    return this.atomCoordinates!.z;
+  };
+
+  get atomTypes(): string[] {
+    if (this._atomTypes === undefined)
+      this._atomTypes = this.parseAtomTypes();
+    return this._atomTypes;
+  }
+
+  // Protected members
+
+  protected file!: string;
+  /** Index running along the string/file being parsed  */
+  protected _atomCount?: number;
+  protected _bondCount?: number;
+  /** The array of X, Y, Z arrays for atomic coordinates */
+  protected atomCoordinates?: CoordinateArrays;
+  protected _atomTypes?: string[];
+
+  protected abstract parseAtomAndBondCounts(): AtomAndBondCounts;
+  protected abstract getCountsLineIdx(): number;
+  /** Get idx of the first line of the atom block containing atom data  */
+  protected abstract getAtomBlockIdx(): number;
+  /** Get idx of the first line of the bond block containing bond data */
+  protected abstract getBondBlockIdx(): number;
+  /** Shift idx from the beginning of the line to X coordinate */
+  protected abstract shiftIdxToXColumn(lineStartIdx: number): number;
+  /** Shift idx from the beginning of the line to atom type column */
+  protected abstract shiftIdxToAtomType(lineStartIdx: number): number;
+  /** Parse atom type at idx */
+  protected abstract parseAtomType(idx: number): string;
+
+  protected setAtomAndBondCounts(): void {
+    const {atomCount, bondCount} = this.parseAtomAndBondCounts();
+    this._atomCount = atomCount;
+    this._bondCount = bondCount;
+  }
+
+  /** Gets the idx of the next column relatively to this._currentIdx */
+  protected getNextColumnIdx(idx: number): number {
+    // skip non-whitespace, if necessary
+    while (!this.isWhitespace(idx))
+      ++idx;
+    // skip whitespace
+    while (this.isWhitespace(idx))
+      ++idx;
+    return idx;
```

Comment:
what's the matter of this method?
at first we skip non-whitespaces, then the whitespaces